pkg/controller/walle/controller.go

```diff
@@ -0,0 +1,435 @@
+			key,
+			nodeInformer,
+		)
+	}
+	return nil
+}
+
+func (wr *WalleController) getKlusterByKey(key string) (*v1.Kluster, error) {
+	o, exists, err := wr.Factories.Kubernikus.Kubernikus().V1().Klusters().Informer().GetIndexer().GetByKey(key)
+	if err != nil {
+		return nil, err
+	}
+	if !exists {
+		return nil, fmt.Errorf("kluster %v was not found", key)
+	}
+	return o.(*v1.Kluster), err
+}
+
+func (wr *WalleController) watchNodesForKluster(kluster *v1.Kluster) error {
+	key, err := cache.MetaNamespaceKeyFunc(kluster)
+	if err != nil {
+		return err
+	}
+
+	i, ok := wr.nodeInformerMap.Load(key)
+	if !ok {
+		return fmt.Errorf("no node informer created for kluster %s/%s", kluster.GetNamespace(), kluster.GetName())
+	}
+
+	informer := i.(*NodeInformer)
+	cache.WaitForCacheSync(
+		wr.stopCh,
+		informer.HasSynced,
+	)
+
+	go func(informer *NodeInformer) {
+		wr.informerWg.Add(1)
+		defer wr.informerWg.Done()
+		for {
+			select {
+			case <-wr.stopCh:
+				informer.close()
+				return
+			default:
+				informer.run()
+			}
+		}
+	}(informer)
+
+	return nil
+}
+
+func (wr *WalleController) klusterAddFunc(obj interface{}) {
+	key, err := cache.MetaNamespaceKeyFunc(obj)
+	if err == nil {
+		wr.queue.Add(key)
+	}
+}
+
+func (wr *WalleController) klusterUpdateFunc(cur, old interface{}) {
+	key, err := cache.MetaNamespaceKeyFunc(cur)
+	if err == nil {
+		wr.queue.Add(key)
+	}
+}
+
+func (wr *WalleController) klusterDeleteFunc(obj interface{}) {
+	key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+	if err == nil {
+		wr.queue.Add(key)
+	}
+	if i, ok := wr.nodeInformerMap.Load(key); ok {
+		i.(*NodeInformer).close()
+		wr.nodeInformerMap.Delete(key)
+	}
+}
+
+func revertKey(key string) (string, string) {
+	s := strings.Split(key, "/")
+	if len(s) == 2 {
+		return s[0], s[1]
+	}
+	return "", ""
+}
+
+// GetStore returns the SharedIndexInformers cache.Store for a given kluster or an error
+func (wr *WalleController) GetStore(kluster *v1.Kluster) (cache.Store, error) {
+	informer, err := wr.getNodeInformerByKluster(kluster)
+	if err != nil {
+		return nil, err
+	}
+	return informer.GetStore(), nil
+}
+
+// GetIndexer returns the SharedIndexInformers cache.Indexer for a given kluster or an error
+func (wr *WalleController) GetIndexer(kluster *v1.Kluster) (cache.Indexer, error) {
+	informer, err := wr.getNodeInformerByKluster(kluster)
+	if err != nil {
+		return nil, err
+	}
+	return informer.GetIndexer(), nil
+}
+
+func (wr *WalleController) getNodeInformerByKluster(kluster *v1.Kluster) (NodeInformer, error) {
+	key, err := cache.MetaNamespaceKeyFunc(kluster)
+	if err != nil {
+		return NodeInformer{}, err
+	}
+	informer, ok := wr.nodeInformerMap.Load(key)
+	if ok {
+		return informer.(NodeInformer), nil
+	}
+	return NodeInformer{}, fmt.Errorf("no informer found for kluster %v", key)
+}
+
+// List returns all Klusters and their SharedIndexInformers
+func (wr *WalleController) List() map[*v1.Kluster]cache.SharedIndexInformer {
```

Comment:
Not sure this method is useful. I don't see a use case for it atm. If we remove it we can also remove `revertKey`

Reply:
removed